src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+   * which were already explicitly removed earlier, or which are already
+   * installed */
+  for (; i < flatpaks_this_commit_wants->len; ++i)
+    {
+      FlatpakToInstall *flatpak_to_install = g_ptr_array_index (flatpaks_this_commit_wants, i);
+
+      if (!g_hash_table_contains (explicitly_removed_flatpaks,
+                                  flatpak_to_install->app_identifier) &&
+          !g_hash_table_contains (already_installed_flatpaks,
+                                  flatpak_to_install->app_identifier))
+        {
+          g_ptr_array_add (resolved_flatpaks,
+                           flatpak_to_install_copy (flatpak_to_install));
+        }
+    }
+
+  return g_steal_pointer (&resolved_flatpaks);
+}
+
+static FlatpakInstallation *
+get_flatpak_installation (GCancellable *cancellable, GError **error)
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
+    {
+      g_autoptr(GFile) override = g_file_new_for_path (override_path);
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static gboolean
+pull_flatpaks (GPtrArray    *flatpaks_to_deploy,
+               GCancellable *cancellable,
+               GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable,
+                                                                          error);
+  gsize i = 0;
+
+  if (!installation)
+    return FALSE;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakToInstall *to_install = g_ptr_array_index (flatpaks_to_deploy, i);
+      g_autoptr(GError) local_error = NULL;
+
+      /* We have to pass in a local_error instance here and check to see
+       * if it was FLATPAK_ERROR_NOT_INSTALLED - this is what will be
+       * thrown if we succeeded at pulling the flatpak into the local
+       * repository but did not deploy it (since no FlatpakInstalledRef
+       * will be returned). */
+      flatpak_installation_install_full (installation,
+                                         FLATPAK_INSTALL_FLAGS_NO_DEPLOY,
+                                         to_install->repo,
+                                         to_install->ref_kind,
+                                         to_install->app_identifier,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         cancellable,
+                                         &local_error);
+
+      /* This only happens if the wrong version of flatpak is deployed
+       * with eos-updater */
+      if (!local_error)
+        {
+          g_critical ("Flatpak installation should not have succeeded!");
```

Comment:
This should return a `GError` and gracefully bail out rather than aborting, since it’s a potential runtime problem rather than a programming problem, right? We do not want the OS updater to abort.

Reply:
In my view, it would be a programmer / deployment error - if the user sets `FLATPAK_INSTALL_FLAGS_NO_PULL` it should be basically impossible for the install to succeed unless the wrong flatpak version was in use. In that sense, this would be "unexpected" behaviour. Perhaps `g_critical` and continuing is the way to go?

Reply:
The guideline I normally go with is that basically every function except `main()` should return `GError`s, and `main()` (or the code which implements the CLI or service interface), and only that, is responsible for logging, printing or ignoring errors.

Reply:
I mean, we *can* bubble up the error here - its just that this case highly highly likely to be a programmer/deployment error as opposed to a runtime error and it may be worth complaining a little more loudly than just using `GError` and expecting the caller to deal with it.

That said, I also understand the merit in not aborting something like the system updater, so perhaps a good approach here would be to both complain loudly and use `GError`. Thoughts?

Reply:
If it’s highly highly unlikely (but still possible to be a runtime error) then I’d bubble it up in a `GError` and not bother complaining any louder about it — it’ll cause the update to fail, and that will get reported loudly enough already.